doc/kernel-limitations.md

```diff
@@ -0,0 +1,174 @@
+| [Memory limitations](#limitations-arisen-from-sgxs-memory-model) | User stack/heap memory is limited to a user setting. Page permissions can be manipulated by host. | Compatibility and Security | Possible with SGX2 |
+| [Clock limitation](#limitations-arisen-from-lack-of-access-to-time-source) | Clock ticks and resolution are controlled by host | Security | Possible but can't be entirely eliminated |
+| [Exception handling limitation](#limitations-arisen-from-handling-of-hardware-exceptions-by-os) | Exception handling has to pass through host  | Performance and security | Possible with SGX2
+| [Network limitation](#network-limitations)  | Networking has to pass through host | Performance and security | Possible but can't be entirely eliminated |
+| [File system limitations](#file-System-limitations) | Support ramfs, hostfs, and ext2 only. Mounting of a file system has to be explicit. Changes are not always persisted. | Compatibility | Possible |
+
+## Limitations arisen from SGX's single process model
+
+With SGX, an enclave application runs as a single process in a single
+address space. Mystikos tries to emulate a process with a thread. When we
+create a child process, we actually create a child thread with proper
+process specific information, including `pid`, `ppid`, `fd table`, etc.
+
+This approach would allow child processes created with `posix_spawn` to
+run alongside the parent process. Sure, there is no memory address space
+separation between the processes, but the child or the parent would have
+no handle to reference the other side's memory.
+
+On the other hand, a child process created with `fork` is supposed to
+share variables with the parent. It's significantly more challenging to
+emulating a forked child process as a thread without underlying support
+of address space separation.
+
+
+## Limitations arisen from SGX's thread model
+
+With SGX1, the number of threads running inside an enclave, i.e.,
+**ethread**, has to be statically declared. Once the application starts,
+no more ethreads can be dynamically created. Meanwhile, the stack size
+of each ethread is also statically declared and allocated while creating
+the enclave. An application attempting to create more ethreads than what
+is declared will crash.
+
+A Mystikos application declares how many user threads are needed in
+config.json. A potential mitigation of the limitation is to make the setting big
+enough. The caveat is that we should ensure the stack size for each ethread
+is reasonable so we don't waste too much stack memory on unused threads.
+
+When an user application calls `pthread_create`, Mystikos makes an OCALL
+to the host to create a host thread, and re-enters the enclave with an
+ECALL, and once inside the enclave, jump to the thread routine. The long
+chain of events leads to delays and performance problems occasionally.
+
+The security implication of using host threads is that the untrusted host
```

Comment:
```suggestion
The host is able to starve or switch execution order of ethreads, but because Mystikos implement synchronization primitives inside the SGX Enclave, the host cannot use the thread scheduling interface to cause unexpected synchronization error as long as the Enclave SW utilizes the synchronization primitives properly.
```

Reply:
@bodzhang Are you referring to the futex primitive inside the enclave? I feel the suggested changes are too broad and too detailed. Can we abstract/scope it somehow?

Reply:
We decided to take this one.